In [1]:
from evaluation import *

ann_comb, ann_SOC, ann_ASOIAF, ann_SA
lengths:  [622, 91, 256, 275]
POVs:  30 9 15 6


In [2]:
from tpot import TPOTClassifier
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split


#### From http://epistasislab.github.io/tpot/using/#scoring-functions :
You can pass the callable object/function with signature scorer(estimator, X, y), where estimator is trained estimator to use for scoring, X are features that will be passed to estimator.predict and y are target values for X. To do this, you should implement your own function. See the example below for further explanation.

In [3]:
def make_scorer(nicknames2name):
    def scorer(estimator, X, y):
        solver = MLCharacterSolver(
            classifier = estimator,
            feature_extractor= lambda x:x, # use identity, as feature extraction is already done, we are not working with raw_texts
            nicknames2name = nicknames2name
        )
        return solver.test(X, y)
    return scorer

# This currently does not work.

In [4]:
def extract_training_cases(ann_data, feature_extractor):
    texts, reference_characters = extract_texts_and_characters(ann_data)
    Xs = [] # Feature vectors
    Ys = [] # Binary as to if this feature is the target
    for reference_name, raw_text in zip(reference_characters, texts):
        names, vectors, _ = feature_extractor(raw_text)
        Ys.extend([(name == reference_name) for name in names])
        Xs.extend(vectors)
        
    return np.asarray(Xs), np.asarray(Ys)

In [5]:
features, labels = extract_training_cases(ann_SOC, hybrid_features)

In [6]:
X_train, X_test, y_train, y_test = train_test_split(features, labels,
                                                    train_size=0.75, test_size=0.25)

tpot = TPOTClassifier(generations=5, population_size=20, verbosity=3, cv=2)#, scoring=make_scorer(nicknames2name_comb))

tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
tpot.export('tpot_pipeline.py')

30 operators have been imported by TPOT.


Optimization Progress:  17%|█▋        | 20/120 [01:37<10:13,  6.14s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 cosine was provided as affinity. Ward can only work with euclidean distances.


Optimization Progress:  33%|███▎      | 40/120 [03:18<06:17,  4.72s/pipeline]

Generation 1 - Current Pareto front scores:
-1	0.9928599673488756	LogisticRegression(input_matrix, LogisticRegression__C=15.0, LogisticRegression__dual=False, LogisticRegression__penalty=l1)
-2	0.9936118470481237	GradientBoostingClassifier(ExtraTreesClassifier(input_matrix, ExtraTreesClassifier__bootstrap=False, ExtraTreesClassifier__criterion=gini, ExtraTreesClassifier__max_features=0.1, ExtraTreesClassifier__min_samples_leaf=19, ExtraTreesClassifier__min_samples_split=19, ExtraTreesClassifier__n_estimators=100), GradientBoostingClassifier__learning_rate=0.1, GradientBoostingClassifier__max_depth=10, GradientBoostingClassifier__max_features=0.2, GradientBoostingClassifier__min_samples_leaf=2, GradientBoostingClassifier__min_samples_split=3, GradientBoostingClassifier__n_estimators=100, GradientBoostingClassifier__subsample=0.1)



Optimization Progress:  33%|███▎      | 40/120 [03:19<06:17,  4.72s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Unsupported set of arguments: The combination of penalty='l1' and loss='squared_hinge' are not supported when dual=True, Parameters: penalty='l1', loss='squared_hinge', dual=True


Optimization Progress:  33%|███▎      | 40/120 [03:20<06:17,  4.72s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Unsupported set of arguments: The combination of penalty='l1' and loss='hinge' is not supported, Parameters: penalty='l1', loss='hinge', dual=False


Optimization Progress:  34%|███▍      | 41/120 [03:21<05:31,  4.19s/pipeline]

Pipeline encountered that has previously been evaluated during the optimization process. Using the score from the previous evaluation.


Optimization Progress:  43%|████▎     | 52/120 [09:23<1:47:18, 94.69s/pipeline]

Skipped pipeline #52 due to time out. Continuing to the next pipeline.


Optimization Progress:  51%|█████     | 61/120 [09:48<09:06,  9.26s/pipeline]  

Generation 2 - Current Pareto front scores:
-1	0.9936109997006038	RandomForestClassifier(input_matrix, RandomForestClassifier__bootstrap=False, RandomForestClassifier__criterion=entropy, RandomForestClassifier__max_features=0.3, RandomForestClassifier__min_samples_leaf=1, RandomForestClassifier__min_samples_split=8, RandomForestClassifier__n_estimators=100)
-2	0.9936118470481237	GradientBoostingClassifier(ExtraTreesClassifier(input_matrix, ExtraTreesClassifier__bootstrap=False, ExtraTreesClassifier__criterion=gini, ExtraTreesClassifier__max_features=0.1, ExtraTreesClassifier__min_samples_leaf=19, ExtraTreesClassifier__min_samples_split=19, ExtraTreesClassifier__n_estimators=100), GradientBoostingClassifier__learning_rate=0.1, GradientBoostingClassifier__max_depth=10, GradientBoostingClassifier__max_features=0.2, GradientBoostingClassifier__min_samples_leaf=2, GradientBoostingClassifier__min_samples_split=3, GradientBoostingClassifier__n_estimators=100, GradientBoostingClassifier__subsa

Optimization Progress:  51%|█████     | 61/120 [09:51<09:06,  9.26s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Input X must be non-negative


Optimization Progress:  51%|█████     | 61/120 [09:53<09:06,  9.26s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Input X must be non-negative


Optimization Progress:  52%|█████▎    | 63/120 [09:54<07:53,  8.31s/pipeline]

Pipeline encountered that has previously been evaluated during the optimization process. Using the score from the previous evaluation.
Pipeline encountered that has previously been evaluated during the optimization process. Using the score from the previous evaluation.


Optimization Progress:  68%|██████▊   | 82/120 [17:02<59:18, 93.64s/pipeline]  

Skipped pipeline #81 due to time out. Continuing to the next pipeline.
Generation 3 - Current Pareto front scores:
-1	0.9936109997006038	RandomForestClassifier(input_matrix, RandomForestClassifier__bootstrap=False, RandomForestClassifier__criterion=entropy, RandomForestClassifier__max_features=0.3, RandomForestClassifier__min_samples_leaf=1, RandomForestClassifier__min_samples_split=8, RandomForestClassifier__n_estimators=100)
-2	0.9936118470481237	GradientBoostingClassifier(ExtraTreesClassifier(input_matrix, ExtraTreesClassifier__bootstrap=False, ExtraTreesClassifier__criterion=gini, ExtraTreesClassifier__max_features=0.1, ExtraTreesClassifier__min_samples_leaf=19, ExtraTreesClassifier__min_samples_split=19, ExtraTreesClassifier__n_estimators=100), GradientBoostingClassifier__learning_rate=0.1, GradientBoostingClassifier__max_depth=10, GradientBoostingClassifier__max_features=0.2, GradientBoostingClassifier__min_samples_leaf=2, GradientBoostingClassifier__min_samples_split=3, Gradient

Optimization Progress:  68%|██████▊   | 82/120 [17:03<41:35, 65.68s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Input X must be non-negative


Optimization Progress:  68%|██████▊   | 82/120 [17:04<41:35, 65.68s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Input X must be non-negative


Optimization Progress:  85%|████████▌ | 102/120 [20:21<02:11,  7.30s/pipeline]

Generation 4 - Current Pareto front scores:
-1	0.9936109997006038	RandomForestClassifier(input_matrix, RandomForestClassifier__bootstrap=False, RandomForestClassifier__criterion=entropy, RandomForestClassifier__max_features=0.3, RandomForestClassifier__min_samples_leaf=1, RandomForestClassifier__min_samples_split=8, RandomForestClassifier__n_estimators=100)
-2	0.9936118470481237	GradientBoostingClassifier(ExtraTreesClassifier(input_matrix, ExtraTreesClassifier__bootstrap=False, ExtraTreesClassifier__criterion=gini, ExtraTreesClassifier__max_features=0.1, ExtraTreesClassifier__min_samples_leaf=19, ExtraTreesClassifier__min_samples_split=19, ExtraTreesClassifier__n_estimators=100), GradientBoostingClassifier__learning_rate=0.1, GradientBoostingClassifier__max_depth=10, GradientBoostingClassifier__max_features=0.2, GradientBoostingClassifier__min_samples_leaf=2, GradientBoostingClassifier__min_samples_split=3, GradientBoostingClassifier__n_estimators=100, GradientBoostingClassifier__subsa

Optimization Progress:  85%|████████▌ | 102/120 [20:25<02:11,  7.30s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Found array with 0 feature(s) (shape=(50, 0)) while a minimum of 1 is required.


Optimization Progress:  85%|████████▌ | 102/120 [20:26<02:11,  7.30s/pipeline]

_pre_test decorator: _random_mutation_operator: num_test=0 Found array with 0 feature(s) (shape=(50, 0)) while a minimum of 1 is required.


Generation 5 - Current Pareto front scores:
-1	0.9936109997006038	RandomForestClassifier(input_matrix, RandomForestClassifier__bootstrap=False, RandomForestClassifier__criterion=entropy, RandomForestClassifier__max_features=0.3, RandomForestClassifier__min_samples_leaf=1, RandomForestClassifier__min_samples_split=8, RandomForestClassifier__n_estimators=100)
-2	0.9936118470481237	GradientBoostingClassifier(ExtraTreesClassifier(input_matrix, ExtraTreesClassifier__bootstrap=False, ExtraTreesClassifier__criterion=gini, ExtraTreesClassifier__max_features=0.1, ExtraTreesClassifier__min_samples_leaf=19, ExtraTreesClassifier__min_samples_split=19, ExtraTreesClassifier__n_estimators=100), GradientBoostingClassifier__learning_rate=0.1, GradientBoostingClassifier__max_depth=10, GradientBoostingClassifier__max_features=0.2, GradientBoostingClassifier__min_samples_leaf=2, GradientBoostingClassifier__min_samples_split=3, GradientBoostingClassifier__n_estimators=100, GradientBoostingClassifier__subsa

True

In [7]:
tpot.export('tpot_pipeline.py')

True

In [8]:
print(tpot.score(X_test, y_test))

0.9887387387387387


In [9]:
tpot.evaluated_individuals_

{'BernoulliNB(RBFSampler(input_matrix, RBFSampler__gamma=0.75), BernoulliNB__alpha=0.01, BernoulliNB__fit_prior=True)': {'crossover_count': 0,
  'generation': 0,
  'internal_cv_score': 0.9740703185461776,
  'mutation_count': 0,
  'operator_count': 2,
  'predecessor': ('ROOT',)},
 'BernoulliNB(RBFSampler(input_matrix, RBFSampler__gamma=0.75), BernoulliNB__alpha=10.0, BernoulliNB__fit_prior=True)': {'crossover_count': 0,
  'generation': 'INVALID',
  'internal_cv_score': 0.976324827847229,
  'mutation_count': 1,
  'operator_count': 2,
  'predecessor': ('BernoulliNB(RBFSampler(input_matrix, RBFSampler__gamma=0.75), BernoulliNB__alpha=0.01, BernoulliNB__fit_prior=True)',)},
 'BernoulliNB(input_matrix, BernoulliNB__alpha=0.01, BernoulliNB__fit_prior=True)': {'crossover_count': 0,
  'generation': 'INVALID',
  'internal_cv_score': 0.9582896007863385,
  'mutation_count': 1,
  'operator_count': 1,
  'predecessor': ('BernoulliNB(RBFSampler(input_matrix, RBFSampler__gamma=0.75), BernoulliNB__alpha


## Classical Features
### ann_SOC
```
exported_pipeline = make_pipeline(
    MaxAbsScaler(),
    LogisticRegression(C=20.0, dual=False, penalty="l2")
)
```

99% on  CV, 98% on held-out test

### ann_ASOIAF
```
exported_pipeline = GradientBoostingClassifier(
    learning_rate=0.1,
    max_depth=5,
    max_features=0.25,
    min_samples_leaf=19,
    min_samples_split=13,
    n_estimators=100,
    subsample=0.55
    )
```
99.58% on CV, 99.6% on held-out test

## Word Emb Features

### ann_soc

```
exported_pipeline = make_pipeline(
    StackingEstimator(estimator=ExtraTreesClassifier(bootstrap=False, criterion="gini", max_features=0.8, min_samples_leaf=11, min_samples_split=8, n_estimators=100)),
    RandomForestClassifier(bootstrap=True, criterion="entropy", max_features=0.55, min_samples_leaf=6, min_samples_split=9, n_estimators=100)
)
```
99.2% on CV, 98.9% on helf-out test

In [10]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import MaxAbsScaler
from sklearn.linear_model import LogisticRegression

exported_pipeline = make_pipeline(
    MaxAbsScaler(),
    LogisticRegression(C=20.0, dual=False, penalty="l2")
)

exported_pipeline.fit(X_train, y_train)
accuracy_score(exported_pipeline.predict(X_test), y_test)

0.9921171171171171